### PR TITLE
filter references for satellite assembly generation down to just the primary assemblies

### LINKED ISSUE #38478: Generating satellite assemblies causes a lot of unnecessary work
### Describe the bug

The `GenerateSatelliteAssembliesForCore` task works today by using the C# compiler to produce satellite images. There is a small amount of code associated with this, mostly a single file with a number of assembly attributes. Because there is code though the compiler must be provided with a set of references and the build simply passes along the reference set for the original project. 

As an individual decision that is quite reasonable but when it's applied to projects of scale it produces a lot of unnecessary work. Consider that even the simplest of .NET Core projects has in excess of 100 references. The average is in the neighborhood of 200 references. That means to compile a satellite assembly the compiler typically parsing 200+ references of which it needs one: corelib. 

This is exacerbated by a number of factors: 

- The satellite assemblies are generated for every target framework 
- Projects typically support 10+ languages 

This can very quickly lead to an incredible amount of unnecessary work in the compiler. 

Consider [dotnet/roslyn][dotnet-roslyn] as an example. The process of generating satellite assemblies causes us to process 172,068 references. The entire build uses 227,765 references so 75% of all references passed to the compiler, and processed by the compiler, are completely unnecessary. Yes the VBCSCompilerServer heavily caches references but it still causes a non-trivial amount of work in the compilr while also polluting the LRU cache with items that aren't actually used. That likely leads to cache inefficiency in later compilations in the build. 

Roslyn is getting here by having a very reasonable setup. There are about ~70 projects (when you iterate through project and their TF) and we support 14 localizations. That quickly adds up to ~1,000 satellite assemblies and with an average of just below 200 references per assembly we get to this incredible number. 

I really think we should reconsider how satellite assembly generation works. We must pass some reference to satellite assemblies for them to build but it does not need to be the full set provided to the parent. It really only needs to be a minimally viable corelib with common assembly level attributes. The `netstandard.dll` alone is more than enough. 

### To Reproduce

Build Roslyn.sln from the [dotnet/roslyn][dotnet-roslyn] repository: 

```cmd
> .\Build.cmd -restore -bl
```

Then examine the binary log and you can track through the csc invocations and their associated reference counts. 

[dotnet-roslyn]: https://github.com/dotnet/roslyn

### PR DESCRIPTION
**Pull request by baronfel**
Fixes #38478.

Satellite assemblies have a very constrained dependency list - essentially just the 'primary assembly' for the runtime platform being compiled for. This means one of mscorlib, netstandard, or System.Runtime. We currently pass the entire set of reference assemblies to the Csc Task when generating satellite assemblies, which means that the Roslyn compiler has to spend quite a bit of time on reference metadata reads, often evicting other more useful assemblies from the Roslyn compiler server's caches.

We can take advantage of this constrained generation domain to limit the set of references passed to the compiler invocation to just the 'primary assemblies', which should save quite a lot of time in aggregate.

## Results

On my machine, for each 'cluster' of CSC invocations for satellite assemblies during the Roslyn build, the first call took ~800ms, and subsequent calls took ~75ms. After this change, the numbers are ~6ms and ~4ms respectively.

Aggregated across a single run of the Roslyn solution on my machine, I get

| Name | Total CoreGenerateSatelliteAssemblies Duration (s) |
| - | - |
| Before | 107.2 |
| After | 66.0 |

<details>
<summary>quick F# Interactive script to grab out the duration of satellite assembly generation from a before.binlog and after.binlog file in the current directory</summary>

```fsharp
#r "nuget: MSBuild.StructuredLogger"

open Microsoft.Build.Logging.StructuredLogger
let beforeLog = Serialization.Read("./before.binlog")
let afterLog = Serialization.Read("./after.binlog")

let getCompilationTimes (b: Build) =
    let satelliteCalls = ResizeArray<_>()

    b.VisitAllChildren<Target> (fun t ->
        if t.Name = "CoreGenerateSatelliteAssemblies" then
            satelliteCalls.Add(t.Duration))

    satelliteCalls
    |> Seq.sumBy (fun d -> d.TotalSeconds)

printfn "Before: %A" (getCompilationTimes beforeLog)
printfn "After: %A" (getCompilationTimes afterLog)
```
</details>

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets [hunk 1]
```diff
@@ -1097,10 +1097,14 @@ Copyright (c) .NET Foundation. All rights reserved.
       <UseSharedCompilation>true</UseSharedCompilation>
     </PropertyGroup>
 
+    <ItemGroup>
+      <SatelliteAssemblyReferences Include="@(ReferencePath)" Condition="'%(Filename)' == 'mscorlib' or '%(Filename)' == 'netstandard' or '%(Filename)' == 'System.Runtime' " />
+    </ItemGroup>
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:1102]: You're correct of course - I should clean up properly!

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets [hunk 2]
```diff
@@ -1097,10 +1097,14 @@ Copyright (c) .NET Foundation. All rights reserved.
       <UseSharedCompilation>true</UseSharedCompilation>
     </PropertyGroup>
 
+    <ItemGroup>
+      <SatelliteAssemblyReferences Include="@(ReferencePath)" Condition="'%(Filename)' == 'mscorlib' or '%(Filename)' == 'netstandard' or '%(Filename)' == 'System.Runtime' " />
```
Comment (jaredpar) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:1101]: Given this is defined and used in the same file, should it be `_SatteliteAssemblyReferences` to indicate it's a _private_ variable? Or is the intent that outside code can depend on this?

Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:1101]: It should definitely be private, I'm just hacking around here.

**Comment by baronfel** [comment]
To test this I need to be able to assert on the values of Items, which the current test framework doesn't support. #38811 tracks expanding the framework to support that. Will come back to this when I can green it up.

**Comment by jaredpar** [comment]
@davidwrighton and @agocke assured me that this approach is runtime approved. What do we need to do in order to hit the submit button here?

**Comment by baronfel** [comment]
It needs to be cleaned up to not leak like @KalleOlaviNiemitalo mentioned, and it needs tests. My main problem is I have a bunch of other work that needs doing, so I probably need to get someone from the team to help get this over the line. @dotnet/domestic-cat maybe this would be a great one to focus on in-between codeflow/triage?

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs [hunk 3]
```diff
@@ -33,13 +33,31 @@ public void It_builds_a_simple_desktop_app()
             });
         }
 
+        [Fact]
```
Comment (jaredpar) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs:36]: Do you have any way to future proof this test? I was thinking essentially something long the lines of 

```csharp
[Theory]
[MemberData(DataStorage.Tfms)]
public void Gandalf_excess_references_to_the_compiler(string tfm
{
        var testProject = new TestProject()
        {
            TargetFrameworks = tfm,
            IsExe = true,
            ProjectSdk = "Microsoft.NET.Sdk"
        };
  ...
}
```

Then in a more central place

```csharp

    public static IEnumerable<object[]> Tfms =>
        new List<object[]>
        {
            new object[] { "net7.0" },
            new object[] { "net8.0" },
        };
```

This way when you release the `net9.0` TFM you update a single place and a number of tests get verified on the new TFM.

Comment (agocke) [reply src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs:36]: Yeah, we built this for the linker/AOT tests. If you just want the current TFM you can do [ToolsetInfo.CurrentTargetFramework](https://github.com/dotnet/sdk/blob/12a7cfd34eb17d8faf8f70e53b20369390f367a3/test/Microsoft.NET.TestFramework/ToolsetInfo.cs#L11)

If you want all supported TFMs you can use [PublishTestTools.SupportedTfms](https://github.com/dotnet/sdk/blob/12a7cfd34eb17d8faf8f70e53b20369390f367a3/test/Microsoft.NET.Publish.Tests/PublishTestUtils.cs#L10). Don't ask me why they're in separate places.

Comment (baronfel) [reply src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs:36]: > Don't ask me why they're in separate places.

Good feedback, will log a refactoring issue for domestic cat to pickup in spare cycles.

**Review by joeloff — APPROVED** [review]

**Review by Forgind — APPROVED** [review]

**Comment by baronfel** [comment]
Quick analysis from an [actual Roslyn insertion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771272&view=results):

| Leg | main branch | this pr |
| - | - | - |
| Windows_Debug_Build | 16:56 | 12:48 |
| Windows_Release_Build | 16:43 | 14:29 |
| Unix_Build | 7:34 | 7:09 |
| Source_Build | 7:34 | 7:47 |

**Pull request merged**